Use secrets for glusterfs provisioning passwords

jsafrane opened the pull request:
- no plain password in StorageClass!
- fix the style along the way
- use PV annotations to pass the configuration from provisioners to deleters, inspired by Ceph RBD provisioning.

~~Proposing 1.4:~~

~~\- GlusterFS provisioning is a new 1.4 feature~~
~~\- if we release GlusterFS provisioner as it is now, we need to support it's API (i.e. plaintext passwords) until 2.0~~
~~\- it can break only GlusterFS provisioning, nothing else~~
~~\- it's easy to revert~~

@kubernetes/sig-storage

fixes #31871

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31869)

<!-- Reviewable:end -->

jsafrane commented:
@k8s-bot ok to test

jsafrane commented:
fixes #31871

rootfs commented on examples/experimental/persistent-volume-provisioning/glusterfs-dp.yaml:
```diff
@@ -6,6 +6,5 @@ provisioner: kubernetes.io/glusterfs
   resturl: "http://127.0.0.1:8081"
-  restauthenabled: "true"
   restuser: "admin"
   restuserkey: "password"
```
this should also be replaced with new param

jsafrane replied:
fixed, added also a sample secret.

rootfs commented on examples/experimental/persistent-volume-provisioning/README.md:
```diff
@@ -94,16 +94,17 @@ provisioner: kubernetes.io/glusterfs
-* `restuserkey` : Gluster REST service user's password which will be used for authentication to the REST server.
+* `resturl` : Gluster REST service url which provision gluster volumes on demand. The format should be a valid URL and this is a mandatory parameter for GlusterFS dynamic provisioner.
+* `userId` : Gluster REST service user who has access to create volumes in the Gluster Trusted Pool. This parameter is optional, empty string will be used when omitted.
+* `secretNamespace` + `secretName` : Identification of Secret instance that containes user password to use when talking to Gluster REST service. These parameters are optional, empty password will be used when both `secretNamespace` and `secretName` are omitted.
```
can you also provide an example of secret and how to create the secret in its namespace as in the example?

jsafrane replied:
link to a secret.yaml added.

humblec replied:
s/containes/contains/

rootfs commented:
Reviewed 3 of 5 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31869)*
<!-- Sent from Reviewable.io -->

rootfs commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.kubernetes.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31869#-:-KQa1Ywr9ig0U5BOnHY7:bnfp4nl)*
<!-- Sent from Reviewable.io -->

rootfs commented:
@humblec

pmorie commented:
@kubernetes/sig-api-machinery

humblec commented on pkg/volume/glusterfs/glusterfs_test.go:
```diff
@@ -236,3 +236,58 @@ func TestPersistentClaimReadOnlyFlag(t *testing.T) {
+		},
+		{
+			// Everything with a value
+			url:             "http://localhost",
```
'url' is actually a combination of server and port as "http://localhost:8081".

jsafrane replied:
the port is optional and this is just an unit test

humblec commented on examples/experimental/persistent-volume-provisioning/glusterfs-dp.yaml:
```diff
@@ -6,6 +6,6 @@ provisioner: kubernetes.io/glusterfs
\ No newline at end of file
+  userId: "admin"
+  secretNamespace: "default"
+  secretName: "heketi-secret"
```
May be this can be renamed to 'gluster-secret' or 'rest-secret' for name consistency among other fields as I chose  generic name for all the other fields.

humblec commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -451,18 +489,19 @@ func (r *glusterfsVolumeProvisioner) Provision() (*api.PersistentVolume, error)
 	sz := int(volume.RoundUpSize(volSizeBytes, 1024*1024*1024))
-	glog.V(2).Infof("glusterfs: create volume of size:%d bytes", volSizeBytes)
-	if p.glusterfsClusterConf.glusterRestUrl == "" {
+	glog.V(2).Infof("glusterfs: create volume of size: %d bytes and configuration %+v", volSizeBytes, p.provisioningConfig)
```
As the provisioningConfig has secretvalue in plaintext, do we need to log it ?

jsafrane commented:
Moving to 1.5

lavalamp commented:
@pmorie why did you tag api machinery sig here?

pmorie commented:
@lavalamp because this establishes a possible precedent for using secrets for cluster-level operations.  I know @deads2k has some opinions about this, thought it would be good to get the rest of that group thinking about this too.  Feel free to delete the tag if you think that I erred.

lavalamp commented:
What does "using secrets for cluster-level operations" mean?

On Thu, Sep 1, 2016 at 11:01 AM, Paul Morie <notifications@github.com>
wrote:

> @lavalamp <https://github.com/lavalamp> because this establishes a
> possible precedent for using secrets for cluster-level operations. I know
> @deads2k <https://github.com/deads2k> has some opinions about this,
> thought it would be good to get the rest of that group thinking about this
> too. Feel free to delete the tag if you think that I erred.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31869#issuecomment-244161099>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAngluXI0pXjdolYtm1fIlWwlyRMJdsRks5qlxMOgaJpZM4Jyl7Y>
> .
>

deads2k commented:
> @lavalamp because this establishes a possible precedent for using secrets for cluster-level operations. I know @deads2k has some opinions about this, thought it would be good to get the rest of that group thinking about this too. Feel free to delete the tag if you think that I erred.

It seems a little weird be referencing a namespaced resource from inside of a cluster-scoped resource.  I can't decide if doing it from an unstructured map makes me more or less happy, but either way it feels like an inversion of my hierarchical expectations.

@liggitt cluster-scoped resource containing a `map[string]string` with two parallel keys which point to a secret's namespace/name and then use a constant key inside that secret.  I think its its avoid having secret information in a cluster-scoped resource.  I get it, but it feel like they really want a cluster-scoped secret.

pmorie commented:
@lavalamp 

> What does "using secrets for cluster-level operations" mean?

In this case, using a secret (which is namespaced) as part of the provisioning of a PV (which is not namespaced)

lavalamp commented:
that does sound wrong to me. Is there a design doc I can read? Why do you
feel the need to do such a thing?

On Thu, Sep 1, 2016 at 1:02 PM, Paul Morie <notifications@github.com> wrote:

> @lavalamp <https://github.com/lavalamp>
>
> What does "using secrets for cluster-level operations" mean?
>
> In this case, using a secret (which is namespaced) as part of the
> provisioning of a PV (which is not namespaced)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31869#issuecomment-244194876>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglsBSwKt1I5TJI5tYZ5m2mpY6YEUKks5qly9HgaJpZM4Jyl7Y>
> .
>

pmorie commented:
@lavalamp I'm convinced now that it is wrong, but it was my idea.  I have another, less-wrong idea now, that I will post about here later (about to leave office).

pmorie commented:
@liggitt and I discussed an idea today that I think can short-circuit this entire discussion.

The priors:

1.  I don't think anyone really likes storing credentials or other secret information in the parameters of storage classes.
2.  Cluster-secrets are a non-starter IMO from an isolation standpoint
3.  The issue with using secrets exists here _because it is for an in-tree provisioner_
4.  Ultimately we want provisioners to live out of tree
5.  Injecting out-of-tree provisioners running in pods with secrets is easy and the same as any other pod

I don't want to go down an expensive, torturous route in order to enable a short-term or medium-term use-case, at least not this expensive and torturous for this particular use-case.  Therefore, I suggest we consider adding configuration options for this to `KubeControllerManagerConfiguration` in `apis/componentconfig`.  We can add first-class options per provisioner type that requires it, or allow provisioners to load a config from a properties-like file (one per provisioner that needs it so we can support all types), and we can bypass this whole mess.

I don't think jumping through the hoops to solve this with secrets is worth it, and this is not what secrets were designed to do **at all**.

Thoughts?  @deads2k @kubernetes/sig-storage

jsafrane commented:
> We can add first-class options per provisioner type that requires it

One volume plugin, can talk to several different servers - e.g. legacy network with two-three deparments, each with its own glusterfs, now being consolidated into one Kubernetes cluster. These will need three different configurations, incl. usernames and passwords.

If the configuration lives outside of StorageClass object, it must be easy to have multiple configurations (map?) and reference the particular configuration from a StorageClass easily (by name?).

> or allow provisioners to load a config from a properties-like file

This was already discussed on storage sig, these files need to be synced to all masters, which makes high availability hard to deploy and keep in sync.

deads2k commented:
> Therefore, I suggest we consider adding configuration options for this to KubeControllerManagerConfiguration in apis/componentconfig. We can add first-class options per provisioner type that requires it, or allow provisioners to load a config from a properties-like file (one per provisioner that needs it so we can support all types), and we can bypass this whole mess.

I think that sounds reasonable.  It would allow you to be more helpful with the known volume plugins, while still having flexibility for future ones that are out of tree.

smarterclayton commented:
Agree to the provisioner config in KCMC.

On Tue, Sep 6, 2016 at 8:01 AM, David Eads <notifications@github.com> wrote:

> Therefore, I suggest we consider adding configuration options for this to
> KubeControllerManagerConfiguration in apis/componentconfig. We can add
> first-class options per provisioner type that requires it, or allow
> provisioners to load a config from a properties-like file (one per
> provisioner that needs it so we can support all types), and we can bypass
> this whole mess.
>
> I think that sounds reasonable. It would allow you to be more helpful with
> the known volume plugins, while still having flexibility for future ones
> that are out of tree.
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31869#issuecomment-244929152>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p5A37hhZtdouzaI6vdK6YC6Al7F5ks5qnVY2gaJpZM4Jyl7Y>
> .
>

eparis commented:
I agree with Paul's assessment that if the provisioners were out of tree we wouldn't be having this problem. We would be able to 'just use secrets' because they would be a pod in a namespace

For in tree I believe we have 2 ways to model things reasonably. Always run provisioners in a 'well known namespace.' I assume this is how out of tree would do it, right? That namespace could be a configuration option to the controller, but I don't love that either to be honest. The other option is to explicitly state where the secret is. A secret is not a name. It is a name/namespace tuple. Any place we have a globally scoped resource referencing a secret by only name we have a problem.

But neither 'forcing it into a well known namespace' or 'better specify the actual thing' solves mount/unmount. For mount/unmount we have a PV and the kubelet. That's it. That PV has to describe a couple of things:

1. where is the thing i'm mounting
2. what 'creds' do I need to mount it

Today we make serious mistakes around this.

1. For gluster we specify the name of a Kind: endpoint in the PV and magically assume that the endpoint is going to exist in the namespace where the bound PVC lives.
2. For ceph rbd it is my understanding that we again specify the name of a Kind: secret in the PV and magically assume that the secret is going to exist in the namespace where the bound PVC lives.

So we basically messed up on both fronts :-(. The suggestion from @pmorie and @smarterclayton  to solve this for 'provisioners' does not translate to mount/unmount.

I personally find the componentconfig suggestion to be an anti-pattern. We are saying 'don't use kubernetes to solve kubernetes problems, do it in something outside via $other_magic.

I don't personally care exactly the mechanism to solve the provisioner problem. But it seems the only way to solve the mount/unmount problem is to make the PV references non-ambiguous. The name of a secret or the name of an endpoint are just completely inadequate. If we accept that these need to be namespace/name tuples in the PV then we may choose to follow that same pattern for the provisioners.

jsafrane commented:
@rootfs AFAIK, `KubeControllerManagerConfiguration` is populated from command line when kube-controller-manager starts and some fields from it are passed to volume plugins during initialization, see https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/plugins.go#L72. After that, the parameters are set to stone and cannot be modified. IMO this is a major drawback.

I can add new GlusterFS related options. It's pretty ugly to pass array of options via command line, so we will support only one Gluster+Heketi. From Heketi point of view, that is probably fine, as it can manage several Gluster servers, I don't know much about Ceph.

rootfs commented:
thanks @jsafrane. I don't feel comfortable of using immutable facilities like `componentconfig`. They create problems:

- once set in immutable facilities, credentials cannot change in `componentconfig`. 
- provisioners that generate dynamic credentials cannot set credentials in `componentconfig`.

erinboyd commented:
@rootfs you make some excellent points. In addition, we should be solving
the issue for secrets in general for storage, not just for provisioners. I
see three issues:
1) passing the secret to a provisioner
2) using the secret to mount/unmount
3) using the secret to pre-create volumes that don't leverage provisioners.

The solution should not provide a way to specify the secret for out-of-tree
vs in-tree. It should be a consistent way it flows through the code.
Ideally we shouldn't be differentiating what type of provisioner it is and
taking some alternate path, then we are maintaining to seperate code bases
around this.



On Wed, Sep 7, 2016 at 7:40 AM, Jan Šafránek <notifications@github.com>
wrote:

> @rootfs <https://github.com/rootfs> AFAIK, KubeControllerManagerConfigura
> tion is populated from command line when kube-controller-manager starts
> and some fields from it are passed to volume plugins during initialization,
> see https://github.com/kubernetes/kubernetes/blob/master/cmd/
> kube-controller-manager/app/plugins.go#L72. After that, the parameters
> are set to stone and cannot be modified. IMO this is a major drawback.
>
> I can add new GlusterFS related options. It's pretty ugly to pass array of
> options via command line, so we will support only one Gluster+Heketi. From
> Heketi point of view, that is probably fine, as it can manage several
> Gluster servers, I don't know much about Ceph.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31869#issuecomment-245283311>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADLTZ5jq5aB8QNEA1SP5CU8eyo0HsexKks5qnr69gaJpZM4Jyl7Y>
> .
>

rootfs commented:
For provisioners use case, I list the use cases for namespaced secrets:

|   | Provision secret   | mount secret for PV  |
|---|---|---|
| in-tree   | currently implemented as name/namespace addressable |  must be accessible via name/namespace |
|  out-of-tree |  either located via name/namespace or embedded in its own Pod |  must be accessible via name/namespace |

humblec commented:
I do think, passing parameters via KCMC and treating those parameters  as 'static entities' is a major drawback.

jsafrane commented:
So, after some discussion with @pmorie @eparis @rootfs @humblec @bc and @erinboyd, we agreed that no solution proposed here is perfect. The best solution is to have external provisioners for GlusterFS and Ceph.

Until we have them, the least bad solution is to have parameters in StorageClass that refer to a Secrets instance:

```
apiVersion: extensions/v1beta1
kind: StorageClass
metadata:
  name: slow
provisioner: kubernetes.io/glusterfs
parameters:
  resturl: "http://127.0.0.1:8081"
  userId: admin
  secretNamespace: kube-system
  secretName: glusterfs
```

All these parameters (`resturl`, `userid`, `secretName`/`Namespace`) are then added to provisioned PVs as annotations, so the StorageClass instance does not need to exists at point 
Brief summary of the other solutions:

* clear username + password in StorageClass.parameters - "too clear" 
* configuration in KubeControllerManagerConfiguration - requires Kubernetes restart to update. Note that Heketi itself (=REST API for provisioning of GlusterFS volumes) will be typically deployed on Kubernetes and admin would need to deploy Kuberneters, run Heketi on it and then restart Kubernetes with Heketi address.


Security co

rootfs commented:
@johscheuer see above comment from @jsafrane

pmorie commented:
Thoughts @deads2k @liggitt

deads2k commented:
> Thoughts @deads2k @liggitt

I'd like to have us move away from using this reference as we move provisioners out of the source tree, but I'm ok with the reference for now.  We should provide guidance about a "standard" namespace to use.  `kube-infra` maybe?

jsafrane committed:
Use secrets for glusterfs provisioning passwords

- no plain password in StorageClass!
- fix the style along the way
- use PV annotations to pass the configuration from provisioners to deleters

jsafrane commented:
rebased + made backward compatible (i.e. restored old `restuser` and `restuserkey` parameters)

jsafrane commented:
@k8s-bot test this issue: #31776

childsb commented:
I reviewed and this PR looks good.  Thanks @humblec @jsafrane and everyone else..
